rustls/src/client/client_conn.rs

```diff
@@ -698,15 +698,42 @@ mod connection {
     }
 
     impl ClientConnection {
-        /// Make a new ClientConnection.  `config` controls how
+        /// Make a new ClientConnection. `config` controls how
         /// we behave in the TLS protocol, `name` is the
         /// name of the server we want to talk to.
         pub fn new(config: Arc<ClientConfig>, name: ServerName<'static>) -> Result<Self, Error> {
+            Self::new_with_alpn(config, name, Vec::new())
```

Comment:
I suggest this passes in `config.alpn_protocols` (ditto below for `UnbufferedClientConnection`). Then we can lose the fall-back code to `config.alpn_protocols` in `client/hs.rs`, which slightly changes the semantics of `new_with_alpn()`:

- `new_with_alpn(.., vec![])` turns off ALPN for this connection irrespective of the `config.alpn_protocols` value
- `new_with_alpn(.., vec![b"hello".to_vec()])` turns it on
- `new_with_alpn(.., config.alpn_protocols.clone())` is possible for callers who conditionally want to use the config's value (equiv to `new(..)`)